### ISSUE TITLE
Question: How to synchronise between instances of CareKit apps?

### ISSUE THREAD

**Issue by advatar** [carekit-apple/CareKit#322]
Not a bug but rather a question. What is the design pattern for synchronising the data of a CareKit app between my own devices or sharing with for instance your doctor. Is OCKSynchronizedStoreManager the right place to inject this if using some other type of non-iCloud solution?

**Comment by erik-apple** [comment]
CareKit does not provide the functionality to synchronize stores across devices or with a cloud service. If you would like to setup some kind of synchronization, you will have to do that yourself.

There are really two parts to this question, synchronizing data from the device to the server, and getting data from the server onto the device.

# Getting Data From The Device To The Server
There are a couple places that you could hook in. Which is best will probably depend on where you are sending/retrieving the data from.

### 1. Subclass OCKSynchronizedStoreManager
The sync manager has some methods on it that are called each time data in the store changes. If you subclass and override these methods, you can use them as an opportunity to send data to your server. This is probably the easies route, but it mixes inter-store synchronization responsibilities into a class already responsible for store<->view synchronization, which isn't very desirable.

```swift
class StoreSynchronizer: OCKSynchronizedStoreManager<OCKStore> {

    open override func store<S>(_ store: S, didAddOutcomes outcomes: [S.Outcome]) where S: OCKStoreProtocol {
        super.store(store, didAddOutcomes: outcomes)

        // Do your extra synchronization work here
        sendNewOutcomesToServer(outcomes)
    }
}
```

### 2. Listen to Notifications From OCKSynchronizedStoreManager
The sync manager also has a Combine publisher on it that will notify you any time something in the store changes. You can setup an object to listen in on changes and respond to them. I like this route more than the one above because the responsibility for inter-store synchronization lives in an object separate from the store synchronizer. Better separation of concerns.

```swift
let storeManager: OCKSynchronizedStoreManager<OCKStore> = //...

let cancellable = storeManager.notificationPublisher.sink { notification in 

    switch notification {

    case let outcomeNotification as OCKOutcomeNotification<OCKStore>:
        handleOutcomeNotification(outcomeNotification)

    case let taskNotification as OCKTaskNotification<OCKStore>:
         handleTaskNotification(taskNotification)
    
    // other cases...
    }
}

func handleOutcomeNotification(_ notification: OCKOutcomeNotification<OCKStore>) {
    switch notification.category {
    case .added: sendNewOutcomesToServer([notification.outcome])
    case .updated: updateOutcomesOnServer([notification.outcome])
    case .deleted: deleteOutcomesFromServer([notification.outcome])
    }
}
```

### 3. Subclass OCKStore
You can go down to an even lower level and subclass `OCKStore`. It has methods on for adding, updating, and deleting all basic entities. You can override those methods and tack on extra processing. This route also mixes responsibilities for two stores into one, but the ease of using this solution may make it a winner for you.

```swift
class MyStore: OCKStore {

    override open func addTasks(_ tasks: [OCKTask], queue: DispatchQueue = .main,
                                completion: ((Result<[OCKTask], OCKStoreError>) -> Void)? = nil) {
        super.addTasks(tasks, queue: queue) { result in 
            switch result {

            case .failure(let error):
                completion(.failure(error))

            case .success(let tasks):
                
                // Your extra sync work goes here
                sendTasksToServer(tasks)

                // You could call completion right away or wait until
                // you hear back from the server.
                completion(tasks)
            }
        }
    }
}
```


**Important caveat**
Some of these options are not possible right now because the properties you need to access are not open/public. These will be opened in an upcoming update. If you need access to any of them now, let us know.

# Getting Data From The Server To The Device
The second half of the problem is getting data from the server onto a device. I can't offer as much advice on this front because it is highly dependent on the capabilities of your back end. You will likely have to setup a class that monitors your server, perhaps using polling or websockets, and call methods against the on-device store when important events occur.

```swift
class ServerWatcher {

    // A reference to the on device store
    let store: OCKStore

    // A method to listen to what's happening on the server
    func handleRemoteServerEvent(_ event: MyServerEvent) {
        switch event {
        
        case .taskDeleted(let task):
            store.deleteTask(task)

        case .taskUpdated(let task):
            store.updatedTas(task)

        // case ... more of the same
        
        }
    }
}
```

Given that your able to put the data into the on device store, CareKit will take care of synchronizing it will the views for you.

# Write Your Own Store
Finally, there is also a way to kill two birds with one store. If you write your own store you can just keep everything directly on the server instead of the device, and then synchronization largely ceases to be a problem. User experience will likely degrade if you don't provide some kind of local caching though, so keep that in mind.

```swift
class ServerStore: OCKStoreProtocol {

    public func fetchTasks(_ anchor: OCKTaskAnchor? = nil, 
                           query: OCKTaskQuery? = nil, 
                           queue: DispatchQueue = .main,
                           completion: @escaping (Result<[OCKTask], OCKStoreError>) -> Void) {

        // Fetch tasks directly from server
        getTasksFromServer(anchor, query) { result in 
            queue.async { completion(result) }
        }
    }

    func addTasks(_ tasks: [OCKTask], queue: DispatchQueue = .main, 
                  completion: ((Result<[OCKTask], OCKStoreError>) -> Void)? = nil) {

        // Save tasks directly to server
        sendTasksToServer(tasks, completion: { result in 
            switch result {

            case .failure(let error):
                queue.async { completion?(.failure(error)) }

            case .success(let tasks):
                queue.async {
                    delegate?.store(self, didAddTasks: tasks)
                    completion?(.success(tasks))
                }
            }
        })
    }
}
```

You can then use this store in place of OCKStore anywhere in your app. Data will seamlessly be pulled down from the server instead of the on-device store.

```swift
let storeManager = OCKSynchronizedStoreManager(wrapping: ServerStore())`
let careViewController = OCKDailyTasksPageViewController(storeMananger: storeManager)
```

**Comment by advatar** [comment]
Thanks for the very extensive answer. I will experiment and report back. The idea is to use a Self Sovereign Identity Hub using Content Addressed Networking from here https://github.com/textileio/ios-textile . That way you get a local node running on device that can synchronise with any number of other nodes.

**Comment by advatar** [comment]
So, I am trying out the last strategy "Write Your Own Store".  It is working out fine. The only thing I am having a problem with is things like localDatabaseID as a URL. This seems to be Core Data specific. What do I need to do here? Would just any  UUID work?

**Comment by erik-apple** [comment]
The `localDatabaseID` can be anything you want it to be, but it should be stable. A UUID would work great. You just need to make sure your store produces the same UUID for a given version of a task every time.

Given two instances of a versioned entity such as an `OCKTask`, we look at the `identifier` to determine if both represent the same task, and we look at the `localDatabaseID` to determine if both of them refer to the same version of that task.

There are some changes coming up that will make writing a store easier. We'll also be removing the requirement to have a `localDatabaseID` soon, which should help you too.

**Comment by erik-apple** [comment]
I'm going to go ahead and close this due to lack of activity. The API changed somewhat between the time that this issue was opened and our stable release, but the principals remain the same. If you need any further clarification, feel free to reopen this issue and continue the conversation.

**Comment by dwatson0101** [comment]
@erik-apple have you exposed the notification publisher?  I can't seem to access it as you listed and it appears the access control is preventing me from getting access to notificationPublisher

```
open class OCKSynchronizedStoreManager:
OCKPatientStoreDelegate, OCKCarePlanStoreDelegate, OCKContactStoreDelegate, OCKTaskStoreDelegate, OCKOutcomeStoreDelegate {

    /// The underlying database.
    public let store: OCKAnyStoreProtocol

    internal lazy var subject = PassthroughSubject<OCKStoreNotification, Never>()
    **internal private (set) lazy var notificationPublisher = subject.share()**
```

**Comment by erik-apple** [comment]
@dwatson0101 We just addressed this in #369!
